### PR TITLE
fix: guard rrweb rebuild against invalid custom element names

### PR DESCRIPTION
**Pull request by pauldambra**
## Summary

- Session replay crashes with `DOMException: "webview" is not a valid custom element name` when a snapshot captured in a context that registered non-spec elements (e.g. Electron's `<webview>`) is replayed in normal Chrome. The recorder flags the element with `isCustom: true`; the replay-side rebuild then calls `customElements.define('webview', ...)` and the synchronous throw aborts the whole rebuild walk.
- Validate the tag name against the HTML PotentialCustomElementName grammar (must contain a hyphen, must start with `[a-z]`) and against the reserved-name set (`annotation-xml`, `color-profile`, `font-face*`, `missing-glyph`) before registering.
- Wrap the `customElements.define` call in try/catch as a belt-and-braces so a future spec edge case can't re-crash replay.

## Test plan

- [x] New parameterised tests in `rebuild.test.ts` cover `webview`, reserved names, and non-hyphen names — none throw, none register.
- [x] Valid names (`my-widget-one`) still register exactly once.
- [x] All 97 `@posthog/rrweb-snapshot` unit tests pass; build clean.
- [ ] Puppeteer-based integration tests require Chrome 131 locally (pre-existing environment issue, unrelated to this change) — will run in CI.

---
*Created with [PostHog Code](https://posthog.com/code?ref=pr)*

### CODE REVIEW

File: packages/rrweb-snapshot/src/rebuild.ts [hunk 1]
```diff
@@ -141,13 +160,17 @@ function buildNode(
...
-            n.tagName,
-            class extends doc.defaultView.HTMLElement {},
-          );
+        ) {
+          try {
+            doc.defaultView.customElements.define(
+              n.tagName,
+              class extends doc.defaultView.HTMLElement {},
+            );
+          } catch {}
```
Comment (pauldambra) [inline packages/rrweb-snapshot/src/rebuild.ts]: > [!NOTE]
> 🤖 Automated comment by **QA Swarm** — not written by a human

**[convergent: qa-team/reliability + paul + xp]** 🟢 LOW

Three reviewers independently flagged the empty `catch {}` — once the validation is in place, this is belt-and-braces without telemetry. Either drop the try/catch (validation is sufficient) or log like the adjacent `adaptCssForReplay` does (`console.warn('Failed to adapt css for replay', error)`), otherwise when someone reports "replay is missing stuff" in six months we've got no breadcrumb that this branch fired. Same ergonomics question as the cross-origin iframe fix in `0b59694c`.

Suggested fix: `console.warn('rrweb: failed to register custom element', n.tagName, e)` inside the catch, optionally gated behind a module-level Set of already-warned names to avoid spam on large snapshots.

File: packages/rrweb-snapshot/src/rebuild.ts [hunk 2]
```diff
@@ -89,6 +89,25 @@ export function createCache(): BuildCache {
...
+  'font-face-name',
+  'missing-glyph',
+]);
+
+const validCustomElementName = /^[a-z][a-z0-9._-]*-[a-z0-9._-]*$/;
+
+function isValidCustomElementName(name: string): boolean {
+  if (typeof name !== 'string' || name.length === 0) return false;
+  if (reservedCustomElementNames.has(name)) return false;
+  return validCustomElementName.test(name);
```
Comment (pauldambra) [inline packages/rrweb-snapshot/src/rebuild.ts]: > [!NOTE]
> 🤖 Automated comment by **QA Swarm** — not written by a human

**[convergent: qa-team/frontend + xp]** 🟢 LOW

The regex `/^[a-z][a-z0-9._-]*-[a-z0-9._-]*$/` narrows to ASCII whereas the WHATWG PotentialCustomElementName grammar allows a broader Unicode range (CJK, emoji, non-ASCII letters per PCENChar). In practice this is almost certainly fine: the recorder only sets `isCustom: true` for elements that `customElements.get()` already resolves, so a false negative here just means the element rebuilds as an unknown `HTMLElement` instead of a registered custom element — functionally indistinguishable for replay, and the try/catch backstops anything the regex lets through.

Two small honesty concerns: (1) the function is named `isValidCustomElementName` but lumps the PCEN shape check with the reserved-name set, and (2) the constant `validCustomElementName` shares a name with the function but has a different contract. Consider renaming the regex constant to `pcenShape` or `customElementNameShape` so a reader doesn't think the regex alone is "valid". Not blocking.

File: packages/rrweb-snapshot/test/rebuild.test.ts [hunk 3]
```diff
@@ -642,4 +642,50 @@ ul li.specified c.\\:hover img {
...
+      ['font-face'],
+      ['annotation-xml'],
+      ['missing-glyph'],
+      ['nohyphen'],
+    ])('does not throw and does not register when tagName is %p', (tagName) => {
+      expect(() => buildCustomElement(tagName)).not.toThrow();
+      expect(window.customElements.get(tagName)).toBeUndefined();
+    });
+
+    it('creates a plain element for invalid custom element names', () => {
```
Comment (pauldambra) [inline packages/rrweb-snapshot/test/rebuild.test.ts]: > [!NOTE]
> 🤖 Automated comment by **QA Swarm** — not written by a human

**[convergent: qa-team/data-integrity + xp]** 🟢 LOW

The parameterised case asserts "does not throw and does not register" but doesn't assert the positive half of the contract — that `buildCustomElement` still returned a usable element for each invalid name. Only the `webview` case gets that coverage in the next `it`; `font-face`, `annotation-xml`, `missing-glyph`, and `nohyphen` don't. One-line fix — add `expect(node).toBeInstanceOf(HTMLElement)` to the `it.each` body so the contract is expressed once for all invalid names.

Optional extras worth considering: `UPPERCASE-TAG`, leading non-letter (`-foo`, `1-foo`), whitespace (`my widget`). Small cost, guards against regex churn. `nohyphen` and `webview` exercise the same "fails the regex" branch today so adding one distinctive case like an uppercase-with-hyphen would widen coverage more than adding another no-hyphen example.

File: packages/rrweb-snapshot/src/rebuild.ts [hunk 4]
```diff
@@ -89,6 +89,25 @@ export function createCache(): BuildCache {
...
+}
```
Comment (pauldambra) [inline packages/rrweb-snapshot/src/rebuild.ts:109]: > [!NOTE]
> 🤖 Automated comment by **QA Swarm** — not written by a human

**[convergent: qa-team/compatibility + paul]** 🟢 LOW — follow-up

Out of scope for a hotfix but worth noting: the recorder (`snapshot.ts` ~L854) sets `isCustom: true` purely based on `customElements.get(tagName)`, which in Electron resolves for built-in non-PCEN names like `webview`. Applying the same `isValidCustomElementName` check on the recorder side would stop propagating bad flags into the event stream, so downstream consumers on older rrweb versions don't hit the same DOMException. This PR's rebuild-side guard protects the posthog-rrweb player but not downstream players. Consider a follow-up.

**Comment by pauldambra** [comment]
> [!NOTE]
> 🤖 Automated comment by **QA Swarm** — not written by a human
>
> Multi-perspective review: qa-team (specialists + generalists), paul-reviewer, xp-reviewer

## Verdict: 💬 APPROVE WITH NITS

Small, well-scoped fix that directly addresses the reported Electron `<webview>` crash and is defence-in-depth via both PCEN/reserved-name validation and the surrounding try/catch. No CRITICAL or HIGH findings; a handful of LOW/NIT suggestions around observability and naming honesty.

### Key findings

- 🟢 **LOW — Empty `catch {}` has no telemetry** (convergent across all three reviewers). Either drop the try/catch or `console.warn` so regressions are visible.
- 🟢 **LOW — Function name / regex constant oversell & undersell the contract**. `isValidCustomElementName` lumps PCEN shape with reserved-name exclusion; the regex is ASCII-only (fine in practice given the recorder path + try/catch backstop).
- 🟢 **LOW — Parameterised test under-expresses the positive half of the contract**. Add `expect(node).toBeInstanceOf(HTMLElement)` so every invalid name asserts a usable element was returned.
- 🟢 **LOW — Follow-up**: apply the same validation on the recorder side (`snapshot.ts`) so bad `isCustom` flags stop entering the event stream for downstream players.

### Convergence

- Empty-catch observability flagged independently by **qa-team/reliability + paul + xp** (3×) — highest confidence item.
- ASCII-only regex + naming honesty flagged by **qa-team/frontend + xp**.
- Test-coverage gap flagged by **qa-team/data-integrity + xp**.
- Recorder-side follow-up flagged by **qa-team/compatibility + paul**.

### Reviewer summaries

| Reviewer | Assessment |
| --- | --- |
| 🔍 qa-team | LOW risk overall — guard is in the right place, minor observability and test-coverage suggestions. |
| 👤 paul | approve with caveats — only real nag is the empty catch with no logging, mirrors the ergonomics question from the cross-origin iframe fix. |
| 📐 xp | Passes simple-design rules: small, correct, well-tested. Quibbles are naming honesty and whether the empty catch is a superfluous part post-validation. |

---
*Automated by QA Swarm — not a human review*

File: packages/rrweb-snapshot/src/rebuild.ts [hunk 5]
```diff
@@ -89,6 +89,25 @@
...
+  'font-face-name',
+  'missing-glyph',
+]);
+
+const validCustomElementName = /^[a-z][a-z0-9._-]*-[a-z0-9._-]*$/;
+
+function isValidCustomElementName(name: string): boolean {
+  if (typeof name !== 'string' || name.length === 0) return false;
+  if (reservedCustomElementNames.has(name)) return false;
+  return validCustomElementName.test(name);
```
[the remark this answers is not in the document]
Comment (pauldambra) [reply packages/rrweb-snapshot/src/rebuild.ts]: > [!NOTE]
> 🤖 Automated comment by **PR Shepherd** — not written by a human

Fixed in `c3c11666` — replaced the polynomial regex with linear imperative checks (non-empty, not reserved, contains a hyphen, starts `[a-z]`). No backtracking anywhere, so ReDoS is no longer possible.

File: packages/rrweb-snapshot/src/rebuild.ts [hunk 6]
```diff
@@ -89,6 +89,25 @@ export function createCache(): BuildCache {
...
+  'color-profile',
+  'font-face',
+  'font-face-src',
+  'font-face-uri',
+  'font-face-format',
+  'font-face-name',
+  'missing-glyph',
+]);
+
+const validCustomElementName = /^[a-z][a-z0-9._-]*-[a-z0-9._-]*$/;
```
[the remark this answers is not in the document]
Comment (pauldambra) [reply packages/rrweb-snapshot/src/rebuild.ts]: > [!NOTE]
> 🤖 Automated comment by **PR Shepherd** — not written by a human

Fixed in `c3c11666` — dropped the ASCII-only regex. The new imperative check only screens out the obvious non-PCEN cases (no hyphen, reserved names, doesn't start `[a-z]`) and defers the full grammar to the browser via the existing try/catch. Spec-valid Unicode names like `math-α` now reach `customElements.define` and are accepted when the browser says so, restoring the previous behaviour for those names.

File: packages/rrweb-snapshot/src/rebuild.ts (hunk 1 above)
Comment (pauldambra) [reply packages/rrweb-snapshot/src/rebuild.ts]: > [!NOTE]
> 🤖 Automated comment by **PR Shepherd** — not written by a human

Fixed in `c3c11666` — the catch now `console.warn`s with the failing tagName, gated behind a module-level `warnedCustomElementNames` Set so a snapshot with many bad names only logs each distinct name once. Matches the `adaptCssForReplay` telemetry pattern mentioned above.

File: packages/rrweb-snapshot/src/rebuild.ts (hunk 2 above)
Comment (pauldambra) [reply packages/rrweb-snapshot/src/rebuild.ts]: > [!NOTE]
> 🤖 Automated comment by **PR Shepherd** — not written by a human

Resolved in `c3c11666` — the regex is gone entirely (replaced with imperative checks to fix the CodeQL ReDoS finding), which incidentally kills the naming-honesty concern too. The function is now `isPlausibleCustomElementName` to acknowledge that it's a cheap pre-flight check and the browser's `define` call is the source of truth.

File: packages/rrweb-snapshot/test/rebuild.test.ts (hunk 3 above)
Comment (pauldambra) [reply packages/rrweb-snapshot/test/rebuild.test.ts]: > [!NOTE]
> 🤖 Automated comment by **PR Shepherd** — not written by a human

Fixed in `c3c11666` — the parameterised case now asserts `toBeInstanceOf(window.HTMLElement)` for every invalid name, so the positive half of the contract ("still returns a usable element") is expressed once across all cases. The existing `webview` test was kept as a focused check that the original tagName is preserved on the element.

File: packages/rrweb-snapshot/src/rebuild.ts (hunk 4 above)
Comment (pauldambra) [reply packages/rrweb-snapshot/src/rebuild.ts:109]: > [!NOTE]
> 🤖 Automated comment by **PR Shepherd** — not written by a human

Acknowledged — keeping this out of scope for the hotfix. Filing a follow-up task to add the same `isPlausibleCustomElementName` screen on the recorder side so bad `isCustom` flags stop entering the event stream for downstream players. Resolving this thread.

**Review by fasyy612 — APPROVED** [review]
LGTM
